src/libmongoc/src/mongoc/mongoc-shared.c

```diff
@@ -0,0 +1,163 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "./mongoc-shared.h"
+
+#include "common-thread-private.h"
+#include <bson/bson.h>
+
+#include <assert.h>
+
+typedef struct _mongoc_shared_ptr_aux {
+   int refcount;
+   void (*dtor) (void *);
+   void *managed;
+} _mongoc_shared_ptr_aux;
+
+static void
+_release_aux (_mongoc_shared_ptr_aux *aux)
+{
+   aux->dtor (aux->managed);
+   bson_free (aux);
+}
+
+static bson_mutex_t g_shared_ptr_spin_mtx = PTHREAD_MUTEX_INITIALIZER;
+
+static void
+_shared_ptr_spin_lock ()
+{
+   bson_mutex_lock (&g_shared_ptr_spin_mtx);
+}
+
+static void
+_shared_ptr_spin_unlock ()
+{
+   bson_mutex_unlock (&g_shared_ptr_spin_mtx);
+}
+
+void
+mongoc_shared_ptr_rebind_raw (mongoc_shared_ptr *const ptr,
+                              void *const pointee,
+                              void (*const dtor) (void *))
+{
+   assert (ptr && "NULL given to mongoc_shared_ptr_rebind_raw()");
+   if (!mongoc_shared_ptr_is_null (*ptr)) {
+      /* Release the old value of the pointer, possibly destroying it */
+      mongoc_shared_ptr_release (ptr);
+   }
+   ptr->ptr = pointee;
+   ptr->_aux = NULL;
+   /* Take the new value */
+   if (pointee != NULL) {
+      assert (dtor != NULL);
+      ptr->_aux = bson_malloc0 (sizeof (_mongoc_shared_ptr_aux));
+      ptr->_aux->dtor = dtor;
+      ptr->_aux->refcount = 1;
+      ptr->_aux->managed = pointee;
+   }
+}
+
+void
+mongoc_shared_ptr_rebind (mongoc_shared_ptr *const out,
+                          mongoc_shared_ptr const from)
+{
+   assert (out &&
+           "NULL given as output argument to mongoc_shared_ptr_rebind()");
+   mongoc_shared_ptr_release (out);
+   *out = mongoc_shared_ptr_take (from);
+}
+
+mongoc_shared_ptr
+mongoc_shared_ptr_create (void *pointee, void (*destroy) (void *))
+{
+   mongoc_shared_ptr ret = {0};
+   mongoc_shared_ptr_rebind_raw (&ret, pointee, destroy);
+   return ret;
+}
+
+void
+mongoc_shared_ptr_rebind_atomic (mongoc_shared_ptr *const out,
+                                 mongoc_shared_ptr const from)
+{
+   struct _mongoc_shared_ptr_aux *prev_aux = NULL;
+   size_t prevcount = 0;
+   assert (out &&
+           "NULL given as output argument to mongoc_shared_ptr_rebind_atomic");
+
+   {
+      _shared_ptr_spin_lock ();
+      prev_aux = out->_aux;
+      if (prev_aux) {
+         prevcount = bson_atomic_int_fetch_sub (
+            &out->_aux->refcount, 1, bson_memorder_seqcst);
+      }
+      *out = from;
+      bson_atomic_int_fetch_add (&out->_aux->refcount, 1, bson_memorder_seqcst);
+      _shared_ptr_spin_unlock ();
+   }
+
+   if (prevcount == 1) {
+      _release_aux (prev_aux);
```

Comment:
Does the previous `out->ptr` need to be freed in this case?

Reply:
Yes, because the rebind-target might have been pointing to something that needs to be released.